src/main/java/dev/jbang/source/resolvers/LiteralScriptResourceResolver.java

```diff
@@ -54,17 +61,39 @@ public String description() {
 		return "Literal stdin";
 	}
 
-	public static ResourceRef stringToResourceRef(String resource, String scriptText) throws IOException {
+	public static ResourceRef stringToResourceRef(String resource, String scriptText,
+			Source.Type forceType) throws IOException {
 		ResourceRef result;
 		String urlHash = Util.getStableID(scriptText);
 		Path cache = Settings.getCacheDir(Cache.CacheClass.stdins).resolve(urlHash);
 		cache.toFile().mkdirs();
 		String basename = urlHash;
-		String suffix = ".jsh";
-		if (hasMainMethod(scriptText)) {
-			suffix = ".java";
-			basename = getMainClass(scriptText).orElse(basename);
+		String suffix;
+
+		if (forceType != null) {
+			// User override wins
+			suffix = "." + forceType.extension;
+			if (forceType == Source.Type.java) {
+				// Only .java needs the class name, .jsh/.kt/... don't care
+				basename = getMainClass(scriptText).orElse(basename);
+			}
+		} else {
+			suffix = ".jsh";
+			if (hasMainMethod(scriptText)) {
+				suffix = ".java";
+				basename = getMainClass(scriptText).orElse(basename);
+			}
+		}
+
+		// Ensure basename is a valid Java identifier for .java files to avoid
```

Comment:
We can remove these comments, it's explained on the `toJavaIdentifier()` method. No need to repeat it.